bin/pasta_maintainers_stats.py

```diff
@@ -0,0 +1,303 @@
+    filename_excluded = dict()
+    filenames_filt = dict()
+    for filename in file_sizes:
+        subsystems = all_maintainers.get_subsystems_by_file(filename)
+        subsystems -= {'THE REST'}
+        loc, byte = file_sizes[filename]
+        for subsystem in subsystems:
+            if subsystem in counts:
+                counts[subsystem].increase(loc, byte,
+                                filter_by_files and filename in filter_by_files)
+                if filename in filter_by_files:
+                    filenames_filt[subsystem] += [filename]
+                else:
+                    filename_excluded[subsystem] += [filename]
+            else:
+                counts[subsystem] = Counter(loc, byte,
+                                filter_by_files and filename in filter_by_files)
+                if filename in filter_by_files:
+                    filenames_filt[subsystem] = [filename]
+                    filename_excluded[subsystem] = []
+                else:
+                    filenames_filt[subsystem] = []
+                    filename_excluded[subsystem] = [filename]
+    if len(filter_by_files) > 0:
+        # If filter is given, only return the relevant lines:
+        counts = {name:item for name,item in counts.items()
+                    if item.display_in_filter()}
+    # Sort subsystems by decreasing total line counts
+    return decreasing_order(counts), filename_excluded, filenames_filt
+
+
+def get_counts_by_maintainer(all_maintainers, file_sizes, filter_by_files = []):
+    maintainer_to_subsystem = dict()
+    # maps maintainers to total file sizes they are responsible for
+    counts = dict()
+    for filename in file_sizes:
+        loc, byte = file_sizes[filename]
+        subsystems = all_maintainers.get_subsystems_by_file(filename)
+        subsystems -= {'THE REST'}
+        for subsystem in subsystems:
+            maintainers = all_maintainers.get_maintainers(subsystem)
+            for maintainer in maintainers[1]:
+                name = pretty_name(maintainer)
+                if name in counts:
+                    maintainer_to_subsystem[name].add(subsystem)
+                    counts[name].increase(loc, byte,
+                                            filename in filter_by_files)
+                else:
+                    maintainer_to_subsystem[name] = set([subsystem])
+                    counts[name] = Counter(loc, byte, 
+                                            filename in filter_by_files)
+
+    if len(filter_by_files) > 0:
+        # If filter is given, only return the relevant lines:
+        counts = {name:item for name,item in counts.items()
+                    if item.display_in_filter()}
+    # Sort maintainers by decreasing total line counts
+    return decreasing_order(counts), maintainer_to_subsystem
+
+
+def get_entries_by_files(all_maintainers, filenames):
+    filenames = sorted(filenames, key=lambda item:os.path.basename(item), 
+                        reverse=True)
```

Comment:
Could you explain what this routine does? I don't unterstand the option --files.

Reply:
Very quickly, it gives you a list of file to entry mapping.

Here is quick dump of the csv output, dumped into Google Sheets and formated and reordered (alphabetically): https://docs.google.com/spreadsheets/d/13O5bstZpENC04EvrTziYKe-DilyLwTH6RQ7sWukS82c/edit?usp=sharing

You can quickly see there is structure and substructure of responsibilities according to the directory etc., maybe we then also find way to visualize that nicely...

Reply:
Thanks, got it. Then I'd like to ask for the following change:

AFAICT, the ordering of the filenames does not matter for the evaluation, right? You alphabetically order filenames for reasons of visualisation.

Could you please remove the ordering from this routine? Ordering is for visualisation and not for processing. So we should, for example, return a dictionary {filename: ([entries], loc, size)}. The caller can then do further postprocessing, like ordering things.
The reason: I was wondering why the order plays a role while iterating over the filenames. If we postpone the ordering, it gets clear that it's just about visualisation.